File: award.rb
class Award
  attr_accessor :name, :expires_in, :quality

  def initialize(name, expires_in, quality)
    @name = name
    @expires_in = expires_in
    @quality = quality
  end

  def update_quality
    return if name == 'Blue Distinction Plus'

    if name == 'Blue First'
      update_blue_first
    elsif name == 'Blue Compare'
      update_blue_compare
    else
      update_normal
    end

    self.expires_in -= 1 unless name == 'Blue Distinction Plus'
    self.quality = [quality, 0].max
    self.quality = [quality, 50].min unless name == 'Blue Distinction Plus'
  end

  private

  def update_blue_first
    return if quality >= 50

    self.quality += 1
    self.quality += 1 if expires_in <= 0
  end

  def update_blue_compare
    return self.quality = 0 if expires_in <= 0

    quality_increase = 1
    quality_increase += 1 if expires_in <= 10
    quality_increase += 1 if expires_in <= 5
    self.quality += quality_increase
  end

  def update_normal
    return if quality <= 0

    quality_degrade = expires_in <= 0 ? 2 : 1
    quality_degrade *= 2 if name == 'Blue Star'
    self.quality -= quality_degrade
  end
end

File: award.py
class Award(object):
    def __init__(self, name=None, expires_in=None, quality=None):
        self.name = name
        self.expires_in = expires_in
        self.quality = quality

    @property
    def quality(self):
        return self._quality

    @quality.setter
    def quality(self, value):
        if self.name == "Blue Distinction Plus":
            self._quality = 80
            return

        if value < 0:
            self._quality = 0
            return

        if value > 50:
            self._quality = 50
            return

        self._quality = value

    def decrement_expires_in(self, amount=1):
        if self.name == "Blue Distinction Plus":
            return
        self.expires_in -= amount
